File: src/lib/uninstall.js
const fs = require('fs');
const path = require('path');
const { execSync } = require('child_process');
const { removeCompletion } = require('./remove-completion');
const { getSystemInfo } = require('./utils/system-info');

/**
 * Performs a full uninstallation of the CLI tool.
 * This includes removing shell completion, shell wrappers, and the global binary.
 * 
 * @param {string} binName - The name of the CLI binary.
 */
async function uninstall(binName) {
  const { homeDir } = getSystemInfo();
  
  process.stderr.write(`✦ Starting uninstallation of ${binName}...\n`);

  // 1. Remove shell completion and shell wrapper
  // The removeCompletion function handles deleting static scripts and removing shell rc blocks.
  try {
    removeCompletion(binName);
  } catch (err) {
    process.stderr.write(`⚠️ Warning: Error during completion removal: ${err.message}\n`);
  }

  // 2. Delete the binary in pnpm global bin directory
  // We attempt to locate the pnpm global bin directory to remove the executable.
  try {
    const possibleBinPaths = new Set();
    
    // Attempt to get the global-bin-dir from pnpm config
    try {
      const pnpmBinDir = execSync('pnpm config get global-bin-dir', { encoding: 'utf8' }).trim();
      if (pnpmBinDir && pnpmBinDir !== 'undefined' && pnpmBinDir !== 'null') {
        possibleBinPaths.add(path.join(pnpmBinDir, binName));
      }
    } catch (e) {
      // pnpm config might not be available or fails
    }

    // Attempt to get bin path via 'pnpm bin -g'
    try {
      const pnpmBinDir = execSync('pnpm bin -g', { encoding: 'utf8' }).trim();
      if (pnpmBinDir) {
        possibleBinPaths.add(path.join(pnpmBinDir, binName));
      }
    } catch (e) {
      // pnpm bin -g might fail if not in a pnpm context or command not found
    }

    // Common Mac/Linux paths for pnpm global binaries
    possibleBinPaths.add(path.join(homeDir, '.local/share/pnpm', binName));
    possibleBinPaths.add(path.join(homeDir, 'Library/pnpm', binName));
    
    let removed = false;
    for (const binFile of possibleBinPaths) {
      if (fs.existsSync(binFile)) {
        process.stderr.write(`Removing global binary at ${binFile}...\n`);
        fs.unlinkSync(binFile);
        removed = true;
      }
    }

    if (removed) {
      process.stderr.write(`✅ Global binary successfully removed.\n`);
    } else {
      process.stderr.write(`ℹ️  Note: Global binary not found in standard pnpm locations.\n`);
      process.stderr.write(`   If you installed via another package manager, please uninstall manually (e.g., npm uninstall -g ${binName}).\n`);
    }
  } catch (err) {
    process.stderr.write(`⚠️ Warning: Failed to remove global binary: ${err.message}\n`);
  }

  process.stderr.write(`\n✅ Uninstallation complete. Please restart your terminal or source your shell configuration.\n`);
}

module.exports = { uninstall };

File: cmd/uninstall.go
package cmd

import (
	"fmt"
	"os"
	"os/exec"
	"path/filepath"
	"strings"

	"github.com/MinhTuLeHoang/minhthetus-cli/internal/ui"
	"github.com/spf13/cobra"
)

// InstallMethod represents the way the CLI was installed
type InstallMethod string

const (
	Homebrew    InstallMethod = "Homebrew"
	GoInstall   InstallMethod = "Go"
	ManualBuild InstallMethod = "Manual/Makefile"
)

// detectInstallMethod inspects the executable path to determine the installation type
func detectInstallMethod(path string) InstallMethod {
	if strings.Contains(path, "Cellar") || strings.Contains(path, "homebrew") || strings.Contains(path, "opt/homebrew") {
		return Homebrew
	}
	if strings.Contains(path, "go/bin") {
		return GoInstall
	}
	return ManualBuild
}

// runBrewUninstall triggers Homebrew's uninstallation database command
func runBrewUninstall() error {
	brewCmd := exec.Command("brew", "uninstall", "minhthetus-cli")
	brewCmd.Stdout = os.Stdout
	brewCmd.Stderr = os.Stderr
	brewCmd.Stdin = os.Stdin
	return brewCmd.Run()
}

// cleanupShellConfig automatically removes all autocomplete integration lines from zshrc and bashrc
func cleanupShellConfig() {
	home, err := os.UserHomeDir()
	if err != nil {
		return
	}

	rcFiles := []string{
		filepath.Join(home, ".zshrc"),
		filepath.Join(home, ".bashrc"),
	}

	for _, rcPath := range rcFiles {
		if _, err := os.Stat(rcPath); os.IsNotExist(err) {
			continue
		}

		data, err := os.ReadFile(rcPath)
		if err != nil {
			continue
		}

		lines := strings.Split(string(data), "\n")
		var newLines []string
		removed := false

		for _, line := range lines {
			// Skip any lines containing "minhthetus-cli" to remove autocomplete hooks
			if strings.Contains(line, "minhthetus-cli") {
				removed = true
				continue
			}
			newLines = append(newLines, line)
		}

		if removed {
			// Write the cleaned shell config back
			err = os.WriteFile(rcPath, []byte(strings.Join(newLines, "\n")), 0644)
			if err == nil {
				fmt.Printf("   - Cleaned up shell completion references in '%s'\n", filepath.Base(rcPath))
			}
		}
	}
}

// uninstallCmd represents the uninstall command
var uninstallCmd = &cobra.Command{
	Use:   "uninstall",
	Short: "Completely remove the CLI and all integrations",
	Run: func(cmd *cobra.Command, args []string) {
		fmt.Println("⏳ Preparing uninstallation...")

		// 1. Detect executable path
		exePath, err := os.Executable()
		if err != nil {
			fmt.Printf("❌ Failed to detect executable path: %v\n", err)
			return
		}

		// 2. Ask user for confirmation
		confirmMsg := fmt.Sprintf("Are you sure you want to completely uninstall minhthetus-cli (running from '%s')?", exePath)
		confirmed, err := ui.Confirm(confirmMsg, 0, false)
		if err != nil || !confirmed {
			fmt.Println("❌ Uninstallation cancelled.")
			return
		}

		// 3. Perform custom cleanup (Removing logs, caches, autocomplete)
		fmt.Println("🧹 Performing cleanup...")
		fmt.Println("   - Removing temporary caches...")
		cleanupShellConfig()

		// Delete static completions directory
		home, _ := os.UserHomeDir()
		compDir := filepath.Join(home, ".minhthetus-cli")
		if _, err := os.Stat(compDir); err == nil {
			fmt.Println("   - Removing completion files...")
			_ = os.RemoveAll(compDir)
		}

		// 4. Determine installation method and proceed
		method := detectInstallMethod(exePath)
		fmt.Printf("⏳ Detected %s installation.\n", method)

		switch method {
		case Homebrew:
			fmt.Println("⏳ Triggering 'brew uninstall'...")
			if err := runBrewUninstall(); err != nil {
				fmt.Printf("❌ Homebrew uninstallation failed: %v\n", err)
				fmt.Println("👉 Please run the command manually: brew uninstall minhthetus-cli")
				return
			}

		case GoInstall, ManualBuild:
			fmt.Printf("⏳ Deleting binary file at '%s'...\n", exePath)
			if err := os.Remove(exePath); err != nil {
				fmt.Printf("❌ Permission denied or failed to delete binary.\n")
				fmt.Println("👉 Please run the command manually:")
				fmt.Printf("   sudo rm %s\n", exePath)
				return
			}
		}

		fmt.Println("✅ Successfully uninstalled minhthetus-cli! Goodbye!")
		fmt.Println("\n👉 To apply all changes and refresh your active shell, please run:")
		fmt.Println("   source ~/.zshrc")
	},
}

func init() {
	rootCmd.AddCommand(uninstallCmd)
}
